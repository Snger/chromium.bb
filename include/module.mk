# Copyright (c) 2012 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

include common.mk

install-header: CC_LIBRARY(src/libevdev.so.0)
	install -D -m 0644 include/libevdev/libevdev.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev.h
	install -D -m 0664 include/libevdev/libevdev_event.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_event.h
	install -D -m 0664 include/libevdev/libevdev_mt.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_mt.h
	install -D -m 0664 include/libevdev/libevdev_log.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_log.h