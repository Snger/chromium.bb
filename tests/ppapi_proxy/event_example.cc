// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// C headers
#include <cassert>
#include <cstdio>

// C++ headers
#include <sstream>
#include <string>

// NaCl
#include "ppapi/c/pp_input_event.h"
#include "ppapi/cpp/instance.h"
#include "ppapi/cpp/module.h"
#include "ppapi/cpp/var.h"

namespace {
const char* const kEventsPropertyName = "events";

// Convert a given modifier to a descriptive string.  Note that the actual
// declared type of modifier in each of the event structs (e.g.,
// PP_InputEvent_Key.modifier) is uint32_t, but it is expected to be
// interpreted as a bitfield of 'or'ed PP_InputEvent_Modifier values.
std::string ModifierToString(uint32_t modifier) {
  std::string s;
  if (modifier & PP_INPUTEVENT_MODIFIER_SHIFTKEY) {
    s += "shift ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CONTROLKEY) {
    s += "ctrl ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ALTKEY) {
    s += "alt ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_METAKEY) {
    s += "meta ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISKEYPAD) {
    s += "keypad ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISAUTOREPEAT) {
    s += "autorepeat ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_LEFTBUTTONDOWN) {
    s += "left-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_MIDDLEBUTTONDOWN) {
    s += "middle-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_RIGHTBUTTONDOWN) {
    s += "right-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CAPSLOCKKEY) {
    s += "caps-lock ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_NUMLOCKKEY) {
    s += "num-lock ";
  }
  return s;
}

std::string MouseButtonToString(PP_InputEvent_MouseButton button) {
  switch (button) {
    case PP_INPUTEVENT_MOUSEBUTTON_NONE:
      return "None";
    case PP_INPUTEVENT_MOUSEBUTTON_LEFT:
      return "Left";
    case PP_INPUTEVENT_MOUSEBUTTON_MIDDLE:
      return "Middle";
    case PP_INPUTEVENT_MOUSEBUTTON_RIGHT:
      return "Right";
    default:
      std::ostringstream stream;
      stream << "Unrecognized ("
             << static_cast<int32_t>(button)
             << ")";
      return stream.str();
  }
}

}  // namespace

class EventInstance : public pp::Instance {
 public:
  explicit EventInstance(PP_Instance instance)
      : pp::Instance(instance) {
    std::printf("EventInstance created.\n");
  }
  virtual ~EventInstance() {}

  void KeyEvent(PP_InputEvent_Key key,
                PP_TimeTicks time,
                const std::string& kind) {
    std::ostringstream stream;
    stream << pp_instance() << ":"
           << " Key event:" << kind.c_str()
           << " modifier:" << ModifierToString(key.modifier)
           << " key_code:" << key.key_code
           << " time:" << time
           << "\n";
    std::printf("%s", stream.str().c_str());
    PostMessage(stream.str());
  }

  void MouseEvent(PP_InputEvent_Mouse mouse_event,
                  PP_TimeTicks time,
                  const std::string& kind) {
    std::ostringstream stream;
    stream << pp_instance() << ":"
           << " Mouse event:" << kind.c_str()
           << " modifier:" << ModifierToString(mouse_event.modifier).c_str()
           << " button:" << MouseButtonToString(mouse_event.button).c_str()
           << " x:" << mouse_event.x
           << " y:" << mouse_event.y
           << " click_count:" << mouse_event.click_count
           << " time:" << time
           << "\n";
    std::printf("%s", stream.str().c_str());
    PostMessage(stream.str());
  }

  void CharEvent(PP_InputEvent_Character char_event, PP_TimeTicks time) {
    std::ostringstream stream;
    stream << pp_instance() << ": Character event."
           << " modifier:" << ModifierToString(char_event.modifier).c_str()
           << " text:" << char_event.text
           << " time:" << time
           << "\n";
    std::printf("%s", stream.str().c_str());
    PostMessage(stream.str());
  }

  void WheelEvent(PP_InputEvent_Wheel wheel_event, PP_TimeTicks time) {
    std::ostringstream stream;
    stream << pp_instance() << ": Wheel event."
           << " modifier:" << ModifierToString(wheel_event.modifier).c_str()
           << " deltax:" << wheel_event.delta_x
           << " deltay:" << wheel_event.delta_y
           << " wheel_ticks_x:" << wheel_event.wheel_ticks_x
           << " wheel_ticks_y:" << wheel_event.wheel_ticks_y
           << " scroll_by_page:"
           << (wheel_event.scroll_by_page ? "true" : "false")
           << "\n";
    std::printf("%s", stream.str().c_str());
    PostMessage(stream.str());
  }

  // Handle an incoming input event by switching on type and dispatching
  // to the appropriate subtype handler.
  virtual bool HandleInputEvent(const PP_InputEvent& event) {
    std::printf("HandleInputEvent called\n");
    switch (event.type) {
      case PP_INPUTEVENT_TYPE_UNDEFINED:
        std::printf("Undefined event.\n");
        break;
      case PP_INPUTEVENT_TYPE_MOUSEDOWN:
        MouseEvent(event.u.mouse, event.time_stamp, "Down");
        break;
      case PP_INPUTEVENT_TYPE_MOUSEUP:
        MouseEvent(event.u.mouse, event.time_stamp, "Up");
        break;
      case PP_INPUTEVENT_TYPE_MOUSEMOVE:
        MouseEvent(event.u.mouse, event.time_stamp, "Move");
        break;
      case PP_INPUTEVENT_TYPE_MOUSEENTER:
        MouseEvent(event.u.mouse, event.time_stamp, "Enter");
        break;
      case PP_INPUTEVENT_TYPE_MOUSELEAVE:
        MouseEvent(event.u.mouse, event.time_stamp, "Leave");
        break;
      case PP_INPUTEVENT_TYPE_MOUSEWHEEL:
        WheelEvent(event.u.wheel, event.time_stamp);
        break;
      case PP_INPUTEVENT_TYPE_RAWKEYDOWN:
        KeyEvent(event.u.key, event.time_stamp, "RawKeyDown");
        break;
      case PP_INPUTEVENT_TYPE_KEYDOWN:
        KeyEvent(event.u.key, event.time_stamp, "Down");
        break;
      case PP_INPUTEVENT_TYPE_KEYUP:
        KeyEvent(event.u.key, event.time_stamp, "Up");
        break;
      case PP_INPUTEVENT_TYPE_CHAR:
        CharEvent(event.u.character, event.time_stamp);
        break;
      default:
        std::printf("Unrecognized event type: %d\n", event.type);
        assert(false);
        return false;
    }
    return true;
  }
};

// The EventModule provides an implementation of pp::Module that creates
// EventInstance objects when invoked.  This is part of the glue code that makes
// our example accessible to ppapi.
class EventModule : public pp::Module {
 public:
  EventModule() : pp::Module() {}
  virtual ~EventModule() {}

  virtual pp::Instance* CreateInstance(PP_Instance instance) {
    std::printf("Creating EventInstance.\n");
    return new EventInstance(instance);
  }
};

// Implement the required pp::CreateModule function that creates our specific
// kind of Module (in this case, EventModule).  This is part of the glue code
// that makes our example accessible to ppapi.
namespace pp {
  Module* CreateModule() {
    std::printf("Creating EventModule.\n");
    return new EventModule();
  }
}
