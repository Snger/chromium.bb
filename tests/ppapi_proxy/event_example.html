<!DOCTYPE html>
<html>
  <!--
  Copyright (c) 2010 The Native Client Authors. All rights reserved.
  Use of this source code is governed by a BSD-style license that can be
  found in the LICENSE file.
  -->
<head>
  <title>PPAPI InputEvent test</title>

  <script type="text/javascript">
    module = null;  // Global application object.
    function moduleDidLoad() {
      module = document.getElementById('event_module');
      // Get the events property to force creation of the instance object when
      // EventInstance::GetInstanceObject is called.
      var initialEventsString = module.events;
    }

    function $(id) {
     return document.getElementById(id);
    }

    function getEvents() {
      try {
        var event_dump = document.getElementById('event_dump');
        event_dump.firstChild.nodeValue = module.events;
      } catch(e) {
        alert(e.message);
      }
    }
  </script>
</head>
<body>
<h1>InputEvent Handling Example</h1>
  <embed name="nacl_module"
         id="event_module"
         width=400 height=400
         nacl="event_example.nmf"
         type="application/x-nacl"
         style="background-color:gray"
         onload="moduleDidLoad();" />
<p>
This example demonstrates input events in ppapi.  Each time an input event
happens in the context of the gray box, the embedded NaCl module prints a
message describing the event to the shell where you started your Browser.
Alternatively, you can click the "Get Events" button to display
all input events that have happened since the last time you clicked the
button.
</p>
<form name="event_form" action="" method="get">
  <input type="button" value="Get Events" onclick="getEvents()"/>
</form>

<h2>Recent Events (press button to update):</h2>
<pre id=event_dump>
no events
</pre>
</body>
</html>
