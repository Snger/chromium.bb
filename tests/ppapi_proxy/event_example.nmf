{
  "program": {
    "x86-32": {"url": "ppapi_event_example_x86-32.nexe"},
    "x86-64": {"url": "ppapi_event_example_x86-64.nexe"},
    "arm": {"url": "ppapi_event_example_arm.nexe"}
  }
}
