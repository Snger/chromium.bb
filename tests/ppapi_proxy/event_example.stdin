# First register a bunch of dummy replays with zero count
# so that those signatures will get registered during "install_upcalls"
# Note: because of the zero count we do not have to get the parameters right
#
# Enumerating all known methods ahead of time helps in many ways:
# * we can delay registering non zero count replays till later which help
#   readability
# * we get an error if for some reason the nexe calls one of these functions.
#   Sadly, the nexe usually fails silently when a method is missing from a
#   service
#
# On the downside, this is currently replicated and must be updated by hand
# when signatures change.

replay 0 HasProperty:CCC:iC *
replay 0 HasMethod:CCC:iC *
replay 0 GetProperty:CCC:CC *
replay 0 GetAllPropertyNames:CC:iCC *
replay 0 SetProperty:CCCC:C *
replay 0 RemoveProperty:CCC:C *
replay 0 Call:CCiCC:CC *
replay 0 Construct:CiCC:CC *
replay 0 Deallocate:C: *
replay 0 PPB_GetInterface:s:i *

replay 0 PPB_Audio_Dev_Create:ii:i *
replay 0 PPB_Audio_Dev_IsAudio:i:i *
replay 0 PPB_Audio_Dev_GetCurrentConfig:i:i *
replay 0 PPB_Audio_Dev_StopPlayback:i:i *
replay 0 PPB_Audio_Dev_StartPlayback:i:i *

replay 0 PPB_AudioConfig_Dev_CreateStereo16Bit:lii:l *
replay 0 PPB_AudioConfig_Dev_IsAudioConfig:l:i *
replay 0 PPB_AudioConfig_Dev_RecommendSampleFrameCount:i:i *
replay 0 PPB_AudioConfig_Dev_GetSampleRate:l:i *
replay 0 PPB_AudioConfig_Dev_GetSampleFrameCount:l:i *

replay 0 PPB_Core_AddRefResource:i: *
replay 0 PPB_Core_ReleaseResource:i: *
replay 0 PPB_Core_GetTime::d *
replay 0 ReleaseResourceMultipleTimes:ii: *

replay 0 PPB_Graphics2D_Create:iCi:i *
replay 0 PPB_Graphics2D_IsGraphics2D:i:i *
replay 0 PPB_Graphics2D_Describe:i:Cii *
replay 0 PPB_Graphics2D_PaintImageData:iiCC: *
replay 0 PPB_Graphics2D_Scroll:iCC: *
replay 0 PPB_Graphics2D_ReplaceContents:ii: *
replay 0 PPB_Graphics2D_Flush:ii:i *

replay 0 PPB_ImageData_GetNativeImageDataFormat::i *
replay 0 PPB_ImageData_IsImageDataFormatSupported:i:i *
replay 0 PPB_ImageData_Create:iiCi:i *
replay 0 PPB_ImageData_IsImageData:i:i *
replay 0 PPB_ImageData_Describe:i:Chii *

replay 0 PPB_Instance_GetWindowObject:i:C *
replay 0 PPB_Instance_GetOwnerElementObject:i:C *
replay 0 PPB_Instance_BindGraphics:ii:i *
replay 0 PPB_Instance_IsFullFrame:i:i *
replay 0 PPB_Instance_ExecuteScript:iCC:CC *

replay 0 PPB_URLRequestInfo_Create:i:l *
replay 0 PPB_URLRequestInfo_IsURLRequestInfo:i:i *
replay 0 PPB_URLRequestInfo_SetProperty:iiC:i *
replay 0 PPB_URLRequestInfo_AppendDataToBody:iC:i *
replay 0 PPB_URLRequestInfo_AppendFileToBody:illld:i *

replay 0 PPB_URLLoader_Open:iii:i *
replay 0 PPB_URLLoader_FollowRedirect:ii:i *
replay 0 PPB_URLLoader_GetUploadProgress:i:lli *
replay 0 PPB_URLLoader_GetDownloadProgress:i:lli *
replay 0 PPB_URLLoader_ReadResponseBody:iii:Ci *
replay 0 PPB_URLLoader_Close:i: *

replay 0 PPB_URLResponseInfo_IsURLResponseInfo:i:i *
replay 0 PPB_URLResponseInfo_GetProperty:ii:C *
replay 0 PPB_URLResponseInfo_GetBodyAsFileRef:i:l *

# define a bunch of variables for readability

set_variable MODULE i(4444)
set_variable INSTANCE i(5555)

# From ppapi/c/pp_input_event.h
#
# struct PP_InputEvent {
#   PP_InputEvent_Type type;
#   int32_t padding;
#   PP_TimeTicks time_stamp;  // double
#   union {
#     struct PP_InputEvent_Key key;
#     struct PP_InputEvent_Character character;
#     struct PP_InputEvent_Mouse mouse;
#     struct PP_InputEvent_Wheel wheel;
#     char padding[64];
#   } u;
# };
#
# struct PP_InputEvent_Mouse {
#   uint32_t modifier;
#   PP_InputEvent_MouseButton button;
#   float x;
#   float y;
#   int32_t click_count;
#  }; // 20 bytes -> 44 bytes paddding
#
#
# struct PP_InputEvent_Key {
#   uint32_t modifier;
#   uint32_t key_code;
# };  // 8 bytes -> 56 bytes padding

set_variable ZERO_44 8:0,8:0,8:0,8:0,8:0,4:0
set_variable ZERO_56 8:0,8:0,8:0,8:0,8:0,8:0,8:0

set_variable TIME_1 8:0x3ff0000000000000
set_variable TIME_2 8:0x3ff1000000000000
set_variable TIME_3 8:0x3ff2000000000000
set_variable TIME_4 8:0x3ff3000000000000
set_variable TIME_5 8:0x3ff4000000000000
set_variable TIME_6 8:0x3ff5000000000000
set_variable TIME_7 8:0x3ff6000000000000

set_variable COORDS 4:0x43840000,4:0x42c80000

set_variable EV_MOUSE_DOWN  R(80,4:0,4:0,${TIME_1},8:64,${COORDS},4:1,${ZERO_44})
set_variable EV_MOUSE_UP    R(80,4:1,4:0,${TIME_2},8:64,${COORDS},4:0,${ZERO_44})

set_variable EV_MOUSE_MOVE  R(80,4:2,4:0,${TIME_3},8:0,${COORDS},4:0,${ZERO_44})

set_variable EV_MOUSE_ENTER R(80,4:3,4:0,${TIME_4},8:0,${COORDS},4:0,${ZERO_44})
set_variable EV_MOUSE_LEAVE R(80,4:4,4:0,${TIME_5},8:0,${COORDS},4:0,${ZERO_44})

set_variable EV_KEY_DOWN    R(80,4:7,4:0,${TIME_6},4:0,4:0x54,${ZERO_56})
set_variable EV_KEY_UP      R(80,4:8,4:0,${TIME_7},4:0,4:0x54,${ZERO_56})

# TODO(robertm) figure out what this represents, maybe PP_Var
set_variable PROPERTY C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)

# TODO(robertm) figure out what this represents, maybe PP_Var
set_variable EXCEPTION R(8,8:0)

# Register  signatures found in all previous replay commands for export
# via install_upcalls
echo
echo "*** REGISTER REPLAY METHODS"
replay_activate

# This is not strictly necessary
echo
echo "*** INITIALIZE PEPPER"
initialize_pepper pepper_desc

# create a upcall service on the channel between nexe and sel_univeral
echo
echo "*** CREATE UPCALL SERVICE"
install_upcalls service_string

echo
echo "*** INIT MODULE"
replay 1 PPB_GetInterface:s:i  s("PPB_Core;0.3") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_Instance;0.4") * i(1)
rpc PPP_InitializeModule i(0) ${MODULE} h(pepper_desc) s("${service_string}") * i(0) i(0)

echo
echo "*** INIT INSTANCE"
set_variable TAGS C(23,id\000name\000src\000style\000type\000)
set_variable VALUES C(74,naclModule\000naclModule\000ppapi.nexe\000background-color:gray\000application/x-nacl\000)
rpc PPP_Instance_DidCreate ${INSTANCE} i(5) ${TAGS} ${VALUES} * i(0)

echo
echo "*** GET INSTANCE OBJECT"
replay 1 PPB_GetInterface:s:i  s("PPB_Var(Deprecated);0.3") * i(1)
rpc PPP_Instance_GetInstanceObject ${INSTANCE} * C(16)
# save result to variable
set_variable OBJECT ${result0}

echo
echo "*** GET PROPERTY (AT THE BEGINNING)"
rpc HasProperty ${OBJECT} ${PROPERTY} ${EXCEPTION} * i(0) C(8)
rpc GetProperty ${OBJECT} ${PROPERTY} ${EXCEPTION} * C(8) C(8)

echo
echo "*** TRIGGER REPAINT"
rpc PPP_Instance_DidChangeView ${INSTANCE} I(4,8,79,400,400) I(4,0,0,400,400) *

echo
echo "*** SENT EVENTS EV_MOUSE_DOWN"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_DOWN} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_UP"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_UP} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_MOVE"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_MOVE} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_ENTER"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_ENTER} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_LEAVE"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_LEAVE} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_UP"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_KEY_UP} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_DOWN"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_KEY_DOWN} * i(0)

echo
echo "*** GET PROPERTY (AT THE END)"
rpc HasProperty ${OBJECT} ${PROPERTY} ${EXCEPTION} * i(0) C(8)
rpc GetProperty ${OBJECT} ${PROPERTY} ${EXCEPTION} * C(1000) C(8)


rpc PPP_ShutdownModule *
