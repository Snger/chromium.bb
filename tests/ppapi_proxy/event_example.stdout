
"*** REPLAY PROLOG"

"*** REGISTER REPLAY METHODS"

"*** INITIALIZE PEPPER"

"*** CREATE UPCALL SERVICE"

"*** INIT MODULE"
rpc call intiated PPP_InitializeModule:iihs:ii
input 0:  i(0)
input 1:  i(4444)
input 2:  h(pepper_desc)
input 3:  @LENGTHY_SERIVE_STRING
Creating EventModule.
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Core;0.3")
output 0:  i(1)
rpc call complete PPP_InitializeModule:iihs:ii


"*** INIT INSTANCE"
rpc call intiated PPP_Instance_DidCreate:iiCC:i
input 0:  i(5555)
input 1:  i(5)
input 2:  C(23,id\x00name\x00src\x00style\x00type\x00)
input 3:  C(74,naclModule\x00naclModule\x00ppapi.nexe\x00background-color:gray\x00application/x-nacl\x00)
Creating EventInstance.
EventInstance created.
rpc call complete PPP_Instance_DidCreate:iiCC:i
output 0:  i(1)


"*** GET INSTANCE OBJECT"
rpc call intiated PPP_Instance_GetInstanceObject:i:C
input 0:  i(5555)
EventInstance::GetInstanceObject
EventHandler created.
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Var(Deprecated\x29;0.3")
output 0:  i(1)
rpc call complete PPP_Instance_GetInstanceObject:i:C


"*** GET PROPERTY (AT THE BEGINNING)"
rpc call intiated HasProperty:CCC:iC
input 0:  @THE_OBJECT
input 1:  C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)
input 2:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete HasProperty:CCC:iC
output 0:  i(1)
output 1:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)

rpc call intiated GetProperty:CCC:CC
input 0:  @THE_OBJECT
input 1:  C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)
input 2:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete GetProperty:CCC:CC
output 0:  C(8,\x05\x00\x00\x00\x00\x00\x00\x00)
output 1:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)


"*** TRIGGER REPAINT"
rpc call intiated PPP_Instance_DidChangeView:iII:
input 0:  i(5555)
input 1:  I(4,8,79,400,400)
input 2:  I(4,0,0,400,400)
rpc call complete PPP_Instance_DidChangeView:iII:


"*** SENT EVENTS EV_MOUSE_DOWN"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Mouse event:Down modifier:left-button-down  button:Left x:264 y:100 click_count:1 time:1
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_UP"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf1?@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Mouse event:Up modifier:left-button-down  button:Left x:264 y:100 click_count:0 time:1.0625
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_MOVE"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf2?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Mouse event:Move modifier: button:Left x:264 y:100 click_count:0 time:1.125
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_ENTER"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf3?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Mouse event:Enter modifier: button:Left x:264 y:100 click_count:0 time:1.1875
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_LEAVE"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf4?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Mouse event:Leave modifier: button:Left x:264 y:100 click_count:0 time:1.25
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_UP"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf6?\x00\x00\x00\x00T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Key event:Up modifier: key_code:84 time:1.375
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_DOWN"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x07\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf5?\x00\x00\x00\x00T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
HandleInputEvent called
5555: Key event:Down modifier: key_code:84 time:1.3125
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** GET PROPERTY (AT THE END)"
rpc call intiated HasProperty:CCC:iC
input 0:  @THE_OBJECT
input 1:  C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)
input 2:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete HasProperty:CCC:iC
output 0:  i(1)
output 1:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)

rpc call intiated GetProperty:CCC:CC
input 0:  @THE_OBJECT
input 1:  C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)
input 2:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete GetProperty:CCC:CC
output 0:  C(560,\x05\x00\x00\x00$\x02\x00\x005555: Mouse event:Down modifier:left-button-down  button:Left x:264 y:100 click_count:1 time:1\n5555: Mouse event:Up modifier:left-button-down  button:Left x:264 y:100 click_count:0 time:1.0625\n5555: Mouse event:Move modifier: button:Left x:264 y:100 click_count:0 time:1.125\n5555: Mouse event:Enter modifier: button:Left x:264 y:100 click_count:0 time:1.1875\n5555: Mouse event:Leave modifier: button:Left x:264 y:100 click_count:0 time:1.25\n5555: Key event:Up modifier: key_code:84 time:1.375\n5555: Key event:Down modifier: key_code:84 time:1.3125\n\x00\x00\x00\x00)
output 1:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)

rpc call intiated PPP_ShutdownModule::
rpc call complete PPP_ShutdownModule::

